File: kernel/net/socket.c
/**
 * @file socket.c
 * @brief Socket API implementation
 */

#include "../include/net/socket.h"
#include "../include/net/udp.h"
#include "../include/net/tcp.h"
#include "../include/net/network.h"
#include "../include/kprintf.h"
#include "../include/debug.h"
#include "../include/mm/heap.h"
#include "../include/sync/spinlock.h"
#include "../include/string.h"

// Socket system state
static struct {
    socket_t* sockets;
    int next_fd;
    spinlock_t lock;
    bool initialized;
} socket_state = {0};

/**
 * Initialize socket system
 */
error_code_t socket_init(void) {
    if (socket_state.initialized) {
        return ERR_OK;
    }
    
    kinfo("Initializing socket system...\n");
    
    socket_state.sockets = NULL;
    socket_state.next_fd = 3;  // Start after stdin, stdout, stderr
    spinlock_init(&socket_state.lock);
    socket_state.initialized = true;
    
    kinfo("Socket system initialized\n");
    return ERR_OK;
}

/**
 * Find socket by file descriptor
 */
static socket_t* socket_find(int fd) {
    spinlock_lock(&socket_state.lock);
    
    socket_t* sock = socket_state.sockets;
    while (sock) {
        if ((int)(sock - socket_state.sockets) == fd - 3) {
            spinlock_unlock(&socket_state.lock);
            return sock;
        }
        sock = sock->next;
    }
    
    spinlock_unlock(&socket_state.lock);
    return NULL;
}

/**
 * Create a socket
 */
int socket_create(int domain, int type, int protocol) {
    if (!socket_state.initialized) {
        socket_init();
    }
    
    if (domain != 2) {  // AF_INET
        return -1;
    }
    
    socket_t* sock = (socket_t*)kmalloc(sizeof(socket_t));
    if (!sock) {
        return -1;
    }
    
    memset(sock, 0, sizeof(socket_t));
    sock->type = type;
    sock->bound = false;
    sock->connected = false;
    
    spinlock_lock(&socket_state.lock);
    sock->next = socket_state.sockets;
    socket_state.sockets = sock;
    int fd = socket_state.next_fd++;
    spinlock_unlock(&socket_state.lock);
    
    return fd;
}

/**
 * Bind socket to address
 */
error_code_t socket_bind(int sockfd, sockaddr_in_t* addr) {
    socket_t* sock = socket_find(sockfd);
    if (!sock || !addr) {
        return ERR_INVALID_ARG;
    }
    
    sock->local_ip = addr->addr;
    sock->local_port = addr->port;
    sock->bound = true;
    
    return ERR_OK;
}

/**
 * Listen on socket (TCP only)
 */
error_code_t socket_listen(int sockfd, int backlog) {
    socket_t* sock = socket_find(sockfd);
    if (!sock) {
        return ERR_INVALID_ARG;
    }
    
    if (sock->type != SOCK_STREAM) {
        return ERR_NOT_SUPPORTED;
    }
    
    if (!sock->bound) {
        return ERR_INVALID_STATE;
    }
    
    // Call TCP listen
    extern error_code_t tcp_listen(uint16_t port);
    return tcp_listen(sock->local_port);
}

/**
 * Connect socket to remote address
 */
error_code_t socket_connect(int sockfd, sockaddr_in_t* addr) {
    socket_t* sock = socket_find(sockfd);
    if (!sock || !addr) {
        return ERR_INVALID_ARG;
    }
    
    sock->remote_ip = addr->addr;
    sock->remote_port = addr->port;
    
    if (sock->type == SOCK_STREAM) {
        // TCP - establish connection
        net_device_t* device = network_find_device("eth0");
        if (!device || !device->up) {
            return ERR_DEVICE_NOT_FOUND;
        }
        
        extern tcp_connection_t* tcp_create_connection(uint32_t local_ip, uint16_t local_port, uint32_t remote_ip, uint16_t remote_port);
        tcp_connection_t* conn = tcp_create_connection(
            device->ip_address, 
            sock->bound ? sock->local_port : 0,
            sock->remote_ip,
            sock->remote_port
        );
        
        if (!conn) {
            return ERR_OUT_OF_MEMORY;
        }
        
        sock->data = conn;
        
        // Send SYN
        extern error_code_t tcp_send_packet(tcp_connection_t* conn, uint8_t flags, void* data, size_t data_len);
        conn->state = TCP_STATE_SYN_SENT;
        tcp_send_packet(conn, TCP_FLAG_SYN, NULL, 0);
        // Simplified - would wait for SYN-ACK in full implementation
        conn->state = TCP_STATE_ESTABLISHED;
    }
    
    sock->connected = true;
    return ERR_OK;
}

/**
 * Accept connection (TCP only)
 */
int socket_accept(int sockfd, sockaddr_in_t* addr) {
    socket_t* sock = socket_find(sockfd);
    if (!sock) {
        return -1;
    }
    
    if (sock->type != SOCK_STREAM) {
        return -1;
    }
    
    if (!sock->bound) {
        return -1;
    }
    
    // Call TCP accept
    extern tcp_connection_t* tcp_accept(uint16_t port);
    tcp_connection_t* conn = tcp_accept(sock->local_port);
    
    if (!conn) {
        return -1; // Would block
    }
    
    // Create new socket for connection
    int new_fd = socket_create(2, SOCK_STREAM, 0); // AF_INET
    if (new_fd < 0) {
        // Close connection if socket creation fails?
        // tcp_close(conn);
        return -1;
    }
    
    socket_t* new_sock = socket_find(new_fd);
    if (new_sock) {
        new_sock->connected = true;
        new_sock->bound = true;
        new_sock->local_ip = conn->local_ip;
        new_sock->local_port = conn->local_port;
        new_sock->remote_ip = conn->remote_ip;
        new_sock->remote_port = conn->remote_port;
        new_sock->data = conn;
        
        if (addr) {
            addr->family = 2; // AF_INET
            addr->port = conn->remote_port;
            addr->addr = conn->remote_ip;
        }
    }
    
    return new_fd;
}

/**
 * Send data on socket
 */
error_code_t socket_send(int sockfd, void* data, size_t len, int flags __attribute__((unused))) {
    socket_t* sock = socket_find(sockfd);
    if (!sock || !data || len == 0) {
        return ERR_INVALID_ARG;
    }
    
    if (!sock->connected && !sock->bound) {
        return ERR_INVALID_STATE;
    }
    
    switch (sock->type) {
        case SOCK_DGRAM: {
            // UDP
            uint32_t dest_ip = sock->connected ? sock->remote_ip : 0;  // Would need destination
            uint16_t dest_port = sock->connected ? sock->remote_port : 0;
            uint16_t src_port = sock->bound ? sock->local_port : 0;
            
            return udp_send(dest_ip, dest_port, src_port, data, len);
        }
        
        case SOCK_STREAM: {
            // TCP
            if (sock->data) {
                extern error_code_t tcp_send(tcp_connection_t* conn, void* data, size_t len);
                return tcp_send((tcp_connection_t*)sock->data, data, len);
            }
            return ERR_INVALID_STATE;
        }
        
        default:
            return ERR_NOT_SUPPORTED;
    }
}

/**
 * Receive data from socket
 */
error_code_t socket_recv(int sockfd, void* buffer, size_t* len, int flags __attribute__((unused))) {
    socket_t* sock = socket_find(sockfd);
    if (!sock || !buffer || !len) {
        return ERR_INVALID_ARG;
    }
    
    switch (sock->type) {
        case SOCK_DGRAM: {
            // UDP
            uint32_t src_ip;
            uint16_t src_port, dest_port;
            
            error_code_t err = udp_receive(buffer, len, &src_ip, &src_port, &dest_port);
            if (err != ERR_OK) {
                return err;
            }
            
            // Check if port matches
            if (sock->bound && dest_port != sock->local_port) {
                return ERR_NOT_FOUND;
            }
            
            return ERR_OK;
        }
        
        case SOCK_STREAM: {
            // TCP
            if (sock->data) {
                extern error_code_t tcp_receive(tcp_connection_t* conn, void* buffer, size_t* len);
                return tcp_receive((tcp_connection_t*)sock->data, buffer, len);
            }
            return ERR_INVALID_STATE;
        }
        
        default:
            return ERR_NOT_SUPPORTED;
    }
}

/**
 * Close socket
 */
error_code_t socket_close(int sockfd) {
    socket_t* sock = socket_find(sockfd);
    if (!sock) {
        return ERR_INVALID_ARG;
    }
    
    spinlock_lock(&socket_state.lock);
    
    // Remove from list
    if (socket_state.sockets == sock) {
        socket_state.sockets = sock->next;
    } else {
        socket_t* prev = socket_state.sockets;
        while (prev && prev->next != sock) {
            prev = prev->next;
        }
        if (prev) {
            prev->next = sock->next;
        }
    }
    
    spinlock_unlock(&socket_state.lock);
    
    // Close TCP connection if exists
    if (sock->type == SOCK_STREAM && sock->data) {
        extern error_code_t tcp_close(tcp_connection_t* conn);
        tcp_close((tcp_connection_t*)sock->data);
        sock->data = NULL;
    }
    
    kfree(sock);
    
    return ERR_OK;
}

/**
 * Set socket option
 */
error_code_t socket_setsockopt(int sockfd, int level, int optname, void* optval, size_t optlen) {
    socket_t* sock = socket_find(sockfd);
    if (!sock || !optval) {
        return ERR_INVALID_ARG;
    }
    
    if (level != 1) {  // SOL_SOCKET
        return ERR_NOT_SUPPORTED;
    }
    
    switch (optname) {
        case 2: {  // SO_REUSEADDR
            if (optlen != sizeof(int)) {
                return ERR_INVALID_ARG;
            }
            return ERR_OK;
        }
        
        case 9: {  // SO_KEEPALIVE
            if (optlen != sizeof(int)) {
                return ERR_INVALID_ARG;
            }
            return ERR_OK;
        }
        
        default:
            return ERR_NOT_SUPPORTED;
    }
}

/**
 * Get socket option
 */
error_code_t socket_getsockopt(int sockfd, int level, int optname, void* optval, size_t* optlen) {
    socket_t* sock = socket_find(sockfd);
    if (!sock || !optval || !optlen) {
        return ERR_INVALID_ARG;
    }
    
    if (level != 1) {  // SOL_SOCKET
        return ERR_NOT_SUPPORTED;
    }
    
    switch (optname) {
        case 2: {  // SO_REUSEADDR
            if (*optlen < sizeof(int)) {
                return ERR_INVALID_ARG;
            }
            *(int*)optval = 0;
            *optlen = sizeof(int);
            return ERR_OK;
        }
        
        case 9: {  // SO_KEEPALIVE
            if (*optlen < sizeof(int)) {
                return ERR_INVALID_ARG;
            }
            *(int*)optval = 0;
            *optlen = sizeof(int);
            return ERR_OK;
        }
        
        default:
            return ERR_NOT_SUPPORTED;
    }
}

File: services/network/src/socket.rs
//! Socket API Implementation
//!
//! Provides BSD-style socket interface for network applications

use crate::tcp;
use crate::udp;
use crate::ip;

/// Socket types
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u32)]
pub enum SocketType {
    Stream = 1,      // TCP
    Datagram = 2,    // UDP
    Raw = 3,         // Raw IP
}

/// Socket address families
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u16)]
pub enum AddressFamily {
    Inet = 2,        // IPv4
    Inet6 = 10,      // IPv6
}

/// Socket state
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SocketState {
    Closed,
    Bound,
    Listening,
    Connected,
    Error,
}

/// Socket address (IPv4)
#[repr(C)]
#[derive(Clone, Copy)]
pub struct SocketAddr {
    pub family: u16,
    pub port: u16,
    pub ip: u32,
    pub zero: [u8; 8],
}

impl SocketAddr {
    pub fn new(ip: u32, port: u16) -> Self {
        Self {
            family: AddressFamily::Inet as u16,
            port: port.to_be(),
            ip: ip.to_be(),
            zero: [0; 8],
        }
    }
}

/// Socket structure
pub struct Socket {
    pub socket_type: SocketType,
    pub state: SocketState,
    pub local_addr: SocketAddr,
    pub remote_addr: SocketAddr,
    pub protocol: u8,
    pub tcp_connection_id: Option<usize>,
    pub receive_buffer: [u8; 65536],
    pub receive_len: usize,
    pub send_buffer: [u8; 65536],
    pub send_len: usize,
}

impl Socket {
    pub fn new(socket_type: SocketType) -> Self {
        Self {
            socket_type,
            state: SocketState::Closed,
            local_addr: SocketAddr::new(0, 0),
            remote_addr: SocketAddr::new(0, 0),
            protocol: match socket_type {
                SocketType::Stream => ip::IP_PROTOCOL_TCP,
                SocketType::Datagram => ip::IP_PROTOCOL_UDP,
                SocketType::Raw => 0,
            },
            tcp_connection_id: None,
            receive_buffer: [0; 65536],
            receive_len: 0,
            send_buffer: [0; 65536],
            send_len: 0,
        }
    }
}

const MAX_SOCKETS: usize = 256;
static mut SOCKETS: [Option<Socket>; MAX_SOCKETS] = [None; MAX_SOCKETS];
static mut SOCKET_COUNT: usize = 0;

/// Create socket
pub fn socket_create(socket_type: SocketType) -> Result<usize, ()> {
    unsafe {
        // Find free socket slot
        for i in 0..MAX_SOCKETS {
            if SOCKETS[i].is_none() {
                SOCKETS[i] = Some(Socket::new(socket_type));
                SOCKET_COUNT += 1;
                return Ok(i);
            }
        }
    }

    Err(()) // No free sockets
}

/// Bind socket to address
pub fn socket_bind(socket_fd: usize, addr: SocketAddr) -> Result<(), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref mut socket) = SOCKETS[socket_fd] {
            if socket.state != SocketState::Closed {
                return Err(()); // Already bound
            }

            socket.local_addr = addr;
            socket.state = SocketState::Bound;

            Ok(())
        } else {
            Err(())
        }
    }
}

/// Listen for connections (TCP only)
pub fn socket_listen(socket_fd: usize, backlog: u32) -> Result<(), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref mut socket) = SOCKETS[socket_fd] {
            if socket.socket_type != SocketType::Stream {
                return Err(()); // Only TCP sockets can listen
            }

            if socket.state != SocketState::Bound {
                return Err(()); // Must be bound first
            }

            // Set up listen queue with backlog size
            // In full implementation, would allocate queue for pending connections
            // For now, just mark as listening
            let _ = backlog;

            socket.state = SocketState::Listening;

            Ok(())
        } else {
            Err(())
        }
    }
}

/// Connect socket to remote address
pub fn socket_connect(socket_fd: usize, addr: SocketAddr) -> Result<(), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref mut socket) = SOCKETS[socket_fd] {
            socket.remote_addr = addr;

            match socket.socket_type {
                SocketType::Stream => {
                    // TCP connect
                    let local_ip = u32::from_be(socket.local_addr.ip);
                    let local_port = u16::from_be(socket.local_addr.port);
                    let remote_ip = u32::from_be(addr.ip);
                    let remote_port = u16::from_be(addr.port);

                    let conn_id = tcp::tcp_create_connection(
                        local_ip,
                        local_port,
                        remote_ip,
                        remote_port,
                    )?;

                    socket.tcp_connection_id = Some(conn_id);

                    // Initiate TCP handshake
                    tcp::tcp_connect(conn_id)?;

                    socket.state = SocketState::Connected;

                    Ok(())
                }
                SocketType::Datagram => {
                    // UDP "connect" just sets remote address
                    socket.state = SocketState::Connected;
                    Ok(())
                }
                SocketType::Raw => {
                    socket.state = SocketState::Connected;
                    Ok(())
                }
            }
        } else {
            Err(())
        }
    }
}

/// Accept incoming connection (TCP only)
pub fn socket_accept(socket_fd: usize) -> Result<(usize, SocketAddr), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref socket) = SOCKETS[socket_fd] {
            if socket.socket_type != SocketType::Stream {
                return Err(());
            }

            if socket.state != SocketState::Listening {
                return Err(());
            }

            // Check listen queue for pending connections
            // In full implementation, would check queue and return next connection
            // For now, return error (no pending connections)
            // Full implementation would:
            // 1. Check if queue has pending connections
            // 2. Create new socket for accepted connection
            // 3. Return new socket FD
            Err(())
        } else {
            Err(())
        }
    }
}

/// Send data on socket
pub fn socket_send(socket_fd: usize, data: &[u8], flags: u32) -> Result<usize, ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref mut socket) = SOCKETS[socket_fd] {
            let _ = flags; // TODO: Handle flags

            match socket.socket_type {
                SocketType::Stream => {
                    // TCP send
                    if let Some(conn_id) = socket.tcp_connection_id {
                        tcp::tcp_send(conn_id, data)?;
                        Ok(data.len())
                    } else {
                        Err(())
                    }
                }
                SocketType::Datagram => {
                    // UDP send
                    let remote_ip = u32::from_be(socket.remote_addr.ip);
                    let remote_port = u16::from_be(socket.remote_addr.port);
                    let local_port = u16::from_be(socket.local_addr.port);

                    udp::udp_send(remote_ip, remote_port, local_port, data)?;
                    Ok(data.len())
                }
                SocketType::Raw => {
                    // Raw IP send
                    let remote_ip = u32::from_be(socket.remote_addr.ip);
                    ip::ip_send(remote_ip, socket.protocol, data)?;
                    Ok(data.len())
                }
            }
        } else {
            Err(())
        }
    }
}

/// Receive data from socket
pub fn socket_recv(socket_fd: usize, buffer: &mut [u8], flags: u32) -> Result<usize, ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref mut socket) = SOCKETS[socket_fd] {
            let _ = flags; // TODO: Handle flags

            match socket.socket_type {
                SocketType::Stream => {
                    // TCP receive
                    if let Some(conn_id) = socket.tcp_connection_id {
                        tcp::tcp_receive(conn_id, buffer)
                    } else {
                        Err(())
                    }
                }
                SocketType::Datagram => {
                    // UDP receive
                    udp::udp_receive(buffer).map(|(len, _, _, _)| len)
                }
                SocketType::Raw => {
                    // Raw IP receive
                    ip::ip_receive(buffer).map(|(len, _, _)| len)
                }
            }
        } else {
            Err(())
        }
    }
}

/// Send data to specific address (UDP)
pub fn socket_sendto(socket_fd: usize, data: &[u8], addr: SocketAddr, flags: u32) -> Result<usize, ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref socket) = SOCKETS[socket_fd] {
            let _ = flags;

            if socket.socket_type != SocketType::Datagram {
                return Err(());
            }

            let remote_ip = u32::from_be(addr.ip);
            let remote_port = u16::from_be(addr.port);
            let local_port = u16::from_be(socket.local_addr.port);

            udp::udp_send(remote_ip, remote_port, local_port, data)?;
            Ok(data.len())
        } else {
            Err(())
        }
    }
}

/// Receive data with source address (UDP)
pub fn socket_recvfrom(socket_fd: usize, buffer: &mut [u8], flags: u32) -> Result<(usize, SocketAddr), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref socket) = SOCKETS[socket_fd] {
            let _ = flags;

            if socket.socket_type != SocketType::Datagram {
                return Err(());
            }

            let (len, src_ip, src_port, _) = udp::udp_receive(buffer)?;
            let addr = SocketAddr::new(src_ip, src_port);

            Ok((len, addr))
        } else {
            Err(())
        }
    }
}

/// Close socket
pub fn socket_close(socket_fd: usize) -> Result<(), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if let Some(ref mut socket) = SOCKETS[socket_fd] {
            // Close TCP connection if present
            if let Some(conn_id) = socket.tcp_connection_id {
                tcp::tcp_close(conn_id)?;
            }

            SOCKETS[socket_fd] = None;
            SOCKET_COUNT -= 1;

            Ok(())
        } else {
            Err(())
        }
    }
}

/// Set socket option
pub fn socket_setsockopt(socket_fd: usize, level: u32, optname: u32, optval: &[u8]) -> Result<(), ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if SOCKETS[socket_fd].is_some() {
            // Implement socket options
            // Common options: SO_REUSEADDR, SO_KEEPALIVE, TCP_NODELAY, etc.
            // For now, just acknowledge (options not fully implemented)
            let _ = (level, optname, optval);
            Ok(())
        } else {
            Err(())
        }
    }
}

/// Get socket option
pub fn socket_getsockopt(socket_fd: usize, level: u32, optname: u32, optval: &mut [u8]) -> Result<usize, ()> {
    unsafe {
        if socket_fd >= MAX_SOCKETS {
            return Err(());
        }

        if SOCKETS[socket_fd].is_some() {
            // Implement socket options
            // Common options: SO_REUSEADDR, SO_KEEPALIVE, TCP_NODELAY, etc.
            // For now, return 0 (options not fully implemented)
            let _ = (level, optname, optval);
            Ok(0)
        } else {
            Err(())
        }
    }
}
